cmd/snap-update-ns/utils.go

```diff
@@ -62,10 +66,88 @@ type ReadOnlyFsError struct {
 	Path string
 }
 
+// Error returns a formatted error message.
 func (e *ReadOnlyFsError) Error() string {
 	return fmt.Sprintf("cannot operate on read-only filesystem at %s", e.Path)
 }
 
+// MimicPath returns the path of the directory where a mimic ought to be constructed.
+func (e *ReadOnlyFsError) MimicPath() string {
+	return e.Path
+}
+
+// TrespassingError is an error when filesystem operation would affect the host.
+type TrespassingError struct {
+	Path string
+}
+
+// MimicPath returns the path of the directory where a mimic ought to be constructed.
+func (e *TrespassingError) Error() string {
+	return fmt.Sprintf("cannot trespass on host filesystem at %s", e.Path)
+}
+
+// MimicPath returns the path of the directory where a mimic ought to be constructed.
+func (e *TrespassingError) MimicPath() string {
+	return e.Path
+}
+
+// TODO: Handle /home/*/snap/* when we do per-user mount namespaces and allow
+// defining layout items that refer to SNAP_USER_DATA and SNAP_USER_COMMON.
+//
+// Note that /snap is "not allowed" because it is handled by existing EROFS code.
+// and thus doesn't need any extra logic.
+var allowedPaths = []string{"/var/snap/", "/tmp/"}
+
+// MimicRequiredError describes errors that require construction of a writable mimic.
+type MimicRequiredError interface {
+	Error() string
+	MimicPath() string
+}
+
+// checkTrespassing returns an error if path cannot be written to without a mimic.
+//
+// Determine if path can be used for writing in a way that does not leak to
+// unexpected places in the host. There are some very well defined places that
+// are always safe to write because snaps are expected to write there:
+// $SNAP_DATA, $SNAP_COMMON and (TBD) per-user variants of those.
+//
+// In addition some places are safe because snap-confine arranges the execution
+// environment so that it is simply safe anyway. Currently this is only /tmp
+// because it is private to the snap and doesn't leak into the host
+func checkTrespassing(fd int, segments []string, i int) error {
+	// If the full path is allowed there's no problem. We don't go and check
+	// the filesystem type because it is just going to spam tests with extra
+	// fstatfs calls and it doesn't change anything.
+	p := "/" + strings.Join(segments, "/") + "/"
+	for _, prefix := range allowedPaths {
+		if strings.HasPrefix(p, prefix) {
+			return nil
+		}
+	}
+
+	// Otherwise it depends on where we are. The filesystem can be either
+	// tmpfs or squashfs. Those are "safe" because tmpfs is usually made by
+	// us for a writable mimic (this is not a perfect check) and squashfs is
```

Comment:
this comment should say something about things like /run /dev/shm etc, afaict they are tmps and they are mounted from outside to inside but then they should be safe because they are prohibited in Validate

Reply:
That's a good point. I think I will move all the lists to a single central module and then have a list of easy-to-eyeball things that have given behaviour.

Reply:
I added extra note below